File: hw5-soln/Agent.cc
// Agent.cc
//
// Fall 2021 HW5 solution.
//
// Written by Larry Holder.

#include <iostream>
#include <list>
#include "Agent.h"
#include "Action.h"

using namespace std;

int MySearchEngine::HeuristicFunction(SearchState* state, SearchState* goalState) {
	int cityBlock = abs(state->location.X - goalState->location.X) + abs(state->location.Y - goalState->location.Y);
	return cityBlock;
	//return 0; // not a good heuristic
}

Agent::Agent ()
{
	// Initialize new agent based on new, unknown world
	worldState.agentLocation = Location(1,1);
	worldState.agentOrientation = RIGHT;
	worldState.agentHasGold = false;
	lastAction = CLIMB; // dummy action
	worldState.worldSize = 3; // HW5: somewhere between 3x3 and 9x9
	worldSizeKnown = false;
	worldState.goldLocation = Location(0,0); // unknown
}

Agent::~Agent ()
{

}

void Agent::Initialize ()
{
	// Initialize agent back to the beginning of the world
	worldState.agentLocation = Location(1,1);
	worldState.agentOrientation = RIGHT;
	worldState.agentAlive = true;
	worldState.agentHasGold = false;
	lastAction = CLIMB; // dummy action
	actionList.clear();
}

Action Agent::Process (Percept& percept)
{
	list<Action> actionList2;
		UpdateState(lastAction, percept);
		if (actionList.empty()) {
			if (percept.Glitter) {
				// HW5.4: If there is gold, then GRAB it
				cout << "Found gold. Grabbing it.\n";
				actionList.push_back(GRAB);
			} else if (worldState.agentHasGold && (worldState.agentLocation == Location(1,1))) {
				// HW5.5: If agent has gold and is in (1,1), then CLIMB
				cout << "Have gold and in (1,1). Climbing.\n";
				actionList.push_back(CLIMB);
			} else if (!worldState.agentHasGold && !(worldState.goldLocation == Location(0,0))) {
				// HW5.6: If agent doesn't have gold, but knows its location, then navigate to that location
				cout << "Moving to known gold location (" << worldState.goldLocation.X << "," << worldState.goldLocation.Y << ").\n";
				actionList2 = searchEngine.FindPath(worldState.agentLocation, worldState.agentOrientation, worldState.goldLocation, worldState.agentOrientation);
				actionList.splice(actionList.end(), actionList2);
			} else if (worldState.agentHasGold && !(worldState.agentLocation == Location(1,1))) {
				// HW5.7: If agent has gold, but isn't in (1,1), then navigate to (1,1)
				cout << "Have gold. Moving to (1,1).\n";
				actionList2 = searchEngine.FindPath(worldState.agentLocation, worldState.agentOrientation, Location(1,1), worldState.agentOrientation);
				actionList.splice(actionList.end(), actionList2);
			} else {
				// HW5.8: If safe unvisited location, then navigate there (should be one)
				Location safeUnvisitedLocation = SafeUnvisitedLocation();
				cout << "Moving to safe unvisited location (" << safeUnvisitedLocation.X << "," << safeUnvisitedLocation.Y << ").\n";
				actionList2 = searchEngine.FindPath(worldState.agentLocation, worldState.agentOrientation, safeUnvisitedLocation, worldState.agentOrientation);
				if (!(actionList2.empty())) {
					actionList.splice(actionList.end(), actionList2);
				} else {
					cout << "ERROR: no path to safe unvisited location\n"; // for debugging
					exit(1);
				}
			}
		}
		Action action = actionList.front();
		actionList.pop_front();
		lastAction = action;
		return action;
}

void Agent::GameOver (int score)
{
	if (score < -1000) {
		// Agent died by going forward into pit or Wumpus
		Percept percept; // dummy, values don't matter
		UpdateState(GOFORWARD, percept, true);
		int X = worldState.agentLocation.X;
		int Y = worldState.agentLocation.Y;
		if (!(MemberLocation(worldState.agentLocation, unsafeLocations))) {
			unsafeLocations.push_back(worldState.agentLocation);
		}
		searchEngine.RemoveSafeLocation(X,Y);
		cout << "Found unsafe location at (" << X << "," << Y << ")\n";
	}
}

void Agent::UpdateState(Action lastAction, Percept& percept, bool gameOver) {
	int X = worldState.agentLocation.X;
	int Y = worldState.agentLocation.Y;
	Orientation orientation = worldState.agentOrientation;

	if (lastAction == TURNLEFT) {
		worldState.agentOrientation = (Orientation) ((orientation + 1) % 4);
	}
	if (lastAction == TURNRIGHT) {
		if (orientation == RIGHT) {
			worldState.agentOrientation = DOWN;
		} else {
			worldState.agentOrientation = (Orientation) (orientation - 1);
		}
	}
	if (lastAction == GOFORWARD) {
		if (percept.Bump) {
			if ((orientation == RIGHT) || (orientation == UP)) {
				cout << "World size known to be " << worldState.worldSize << "x" << worldState.worldSize << endl;
				worldSizeKnown = true;
				RemoveOutsideLocations();
			}
		} else {
			switch (orientation) {
			case UP:
				worldState.agentLocation.Y = Y + 1;
				break;
			case DOWN:
				worldState.agentLocation.Y = Y - 1;
				break;
			case LEFT:
				worldState.agentLocation.X = X - 1;
				break;
			case RIGHT:
				worldState.agentLocation.X = X + 1;
				break;
			}
		}
	}
	if (lastAction == GRAB) { // Assume GRAB only done if Glitter was present
		worldState.agentHasGold = true;
	}
	if (lastAction == CLIMB) {
		// do nothing; if CLIMB worked, this won't be executed anyway
	}
	// HW5 requirement 3a
	if (percept.Glitter) {
		worldState.goldLocation = worldState.agentLocation;
		cout << "Found gold at (" << worldState.goldLocation.X << "," << worldState.goldLocation.Y << ")\n";
	}
	// HW5 clarification: track world size
	int new_max = max(worldState.agentLocation.X, worldState.agentLocation.Y);
	if (new_max > worldState.worldSize) {
		worldState.worldSize = new_max;
	}
	// HW5 requirement 3b
	if (!gameOver) {
		UpdateSafeLocations(worldState.agentLocation);
	}
	// HW5 requirement 3c
	if (!(MemberLocation(worldState.agentLocation, visitedLocations))) {
		visitedLocations.push_back(worldState.agentLocation);
	}
}

bool Agent::MemberLocation(Location& location, list<Location>& locationList) {
	if (find(locationList.begin(), locationList.end(), location) != locationList.end()) {
		return true;
	}
	return false;
}

Location Agent::SafeUnvisitedLocation() {
	// Find and return safe unvisited location.
	list<Location>::iterator loc_itr;
	for (loc_itr = safeLocations.begin(); loc_itr != safeLocations.end(); ++loc_itr) {
		if (!(MemberLocation(*loc_itr, visitedLocations))) {
			return *loc_itr;
		}
	}
	return Location(0,0);
}

void Agent::UpdateSafeLocations(Location& location) {
	// HW5 requirement 3b, and HW5 clarification about not known to be unsafe locations
    // Add current and adjacent locations to safe locations, unless known to be unsafe.
	if (!(MemberLocation(location, safeLocations))) {
		safeLocations.push_back(location);
		searchEngine.AddSafeLocation(location.X, location.Y);
	}
	list<Location> adj_locs;
	AdjacentLocations(location, adj_locs);
	list<Location>::iterator loc_itr;
	for (loc_itr = adj_locs.begin(); loc_itr != adj_locs.end(); ++loc_itr) {
		if ((!(MemberLocation(*loc_itr, safeLocations))) && (!(MemberLocation(*loc_itr, unsafeLocations)))) {
			safeLocations.push_back(*loc_itr);
			searchEngine.AddSafeLocation(loc_itr->X, loc_itr->Y);
		}
	}
}

void Agent::RemoveOutsideLocations() {
	// Know exact world size, so remove locations outside the world.
	int boundary = worldState.worldSize + 1;
	for (int i = 1; i < boundary; ++i) {
		safeLocations.remove(Location(i,boundary));
		searchEngine.RemoveSafeLocation(i,boundary);
		safeLocations.remove(Location(boundary,i));
		searchEngine.RemoveSafeLocation(boundary,i);
	}
	safeLocations.remove(Location(boundary,boundary));
	searchEngine.RemoveSafeLocation(boundary,boundary);
}

void Agent::AdjacentLocations(Location& location, list<Location>& adjacentLocations) {
	// Append locations adjacent to given location on to give locations list.
	// One row/col beyond unknown world size is okay. Locations outside the world
	// will be removed later.
	int X = location.X;
	int Y = location.Y;
	if (X > 1) {
		adjacentLocations.push_back(Location(X-1,Y));
	}
	if (Y > 1) {
		adjacentLocations.push_back(Location(X,Y-1));
	}
	if (worldSizeKnown) {
		if (X < worldState.worldSize) {
			adjacentLocations.push_back(Location(X+1,Y));
		}
		if (Y < worldState.worldSize) {
			adjacentLocations.push_back(Location(X,Y+1));
		}
	} else {
		adjacentLocations.push_back(Location(X+1,Y));
		adjacentLocations.push_back(Location(X,Y+1));
	}
}

File: hw5-soln/Agent.py
# Agent.py
#
# Fall 2021 HW5 solution.
#
# Written by Larry Holder.

import Action
import Percept
import Orientation
import Search

class MySearchEngine(Search.SearchEngine):
    def HeuristicFunction(self, state, goalState):
        city_block = abs(state.location[0] - goalState.location[0]) + abs(state.location[1] - goalState.location[1])
        return city_block 
        #return 0 # not a good heuristic
    
class Agent:
    def __init__(self):
        # Initialize new agent based on new, unknown world
        self.agentLocation = [1,1]
        self.agentOrientation = Orientation.RIGHT
        self.agentHasGold = False
        self.lastAction = Action.CLIMB # dummy action
        self.actionList = []
        self.searchEngine = MySearchEngine()
        self.worldSize = 3 # HW5: somewhere between 3x3 and 9x9
        self.worldSizeKnown = False 
        self.goldLocation = [0,0] # unknown
        self.visitedLocations = []
        self.safeLocations = [] # For HW5, means not known to be unsafe
        self.unsafeLocations = []
    
    def __del__(self):
        pass
    
    def Initialize(self):
        # Initialize agent back to the beginning of the world
        self.agentLocation = [1,1]
        self.agentOrientation = Orientation.RIGHT
        self.agentHasGold = False
        self.lastAction = Action.CLIMB # dummy action
        self.actionList = []
    
    # Input percept is a dictionary [perceptName: boolean]
    def Process (self, percept):
        actionList2 = []
        self.UpdateState(self.lastAction, percept)
        if (self.actionList == []):
            if (percept.glitter):
                # HW5.4: If there is gold, then GRAB it
                print("Found gold. Grabbing it.")
                self.actionList.append(Action.GRAB)
            elif (self.agentHasGold and (self.agentLocation == [1,1])):
                # HW5.5: If agent has gold and is in (1,1), then CLIMB
                print("Have gold and in (1,1). Climbing.")
                self.actionList.append(Action.CLIMB)
            elif ((not self.agentHasGold) and (self.goldLocation != [0,0])):
                # HW5.6: If agent doesn't have gold, but knows its location, then navigate to that location
                print("Moving to known gold location (" + str(self.goldLocation[0]) + "," + str(self.goldLocation[1]) + ").")
                actionList2 = self.searchEngine.FindPath(self.agentLocation, self.agentOrientation,
                                                         self.goldLocation, self.agentOrientation)
                self.actionList.extend(actionList2)
            elif (self.agentHasGold and (self.agentLocation != [1,1])):
                # HW5.7: If agent has gold, but isn't in (1,1), then navigate to (1,1)
                print("Have gold. Moving to (1,1).")
                actionList2 = self.searchEngine.FindPath(self.agentLocation, self.agentOrientation,
                                                         [1,1], self.agentOrientation)
                self.actionList.extend(actionList2)
            else:
                # HW5.8: If safe unvisited location, then navigate there (should be one)
                safeUnvisitedLocation = self.SafeUnvisitedLocation()
                print("Moving to safe unvisited location " + str(safeUnvisitedLocation))
                actionList2 = self.searchEngine.FindPath(self.agentLocation, self.agentOrientation,
                                                         safeUnvisitedLocation, self.agentOrientation)
                if actionList2:
                    self.actionList.extend(actionList2)
                else:
                    print("ERROR: no path to safe unvisited location") # for debugging
                    sys.exit(1)
        action = self.actionList.pop(0)
        self.lastAction = action
        return action
    
    def GameOver(self, score):
        if score < -1000:
            # Agent died by going forward into pit or Wumpus
            percept = Percept.Percept() # dummy, values don't matter
            self.UpdateState(Action.GOFORWARD, percept, game_over=True)
            location = self.agentLocation
            if location not in self.unsafeLocations:
                self.unsafeLocations.append(location)
            self.searchEngine.RemoveSafeLocation(location[0], location[1])
            print("Found unsafe location " + str(location))
        return
    
    def UpdateState(self, lastAction, percept, game_over=False):
        X = self.agentLocation[0]
        Y = self.agentLocation[1]
        orientation = self.agentOrientation
        if (lastAction == Action.TURNLEFT):
            self.agentOrientation = (orientation + 1) % 4
        if (lastAction == Action.TURNRIGHT):
            if (orientation == Orientation.RIGHT):
                self.agentOrientation = Orientation.DOWN
            else:
                self.agentOrientation = orientation - 1
        if (lastAction == Action.GOFORWARD):
            if percept.bump:
                if (orientation == Orientation.RIGHT) or (orientation == Orientation.UP):
                    print("World size known to be " + str(self.worldSize) + "x" + str(self.worldSize))
                    self.worldSizeKnown = True
                    self.RemoveOutsideLocations()
            else:
                if orientation == Orientation.UP:
                    self.agentLocation = [X,Y+1]
                elif orientation == Orientation.DOWN:
                    self.agentLocation = [X,Y-1]
                elif orientation == Orientation.LEFT:
                    self.agentLocation = [X-1,Y]
                elif orientation == Orientation.RIGHT:
                    self.agentLocation = [X+1,Y]
        if (lastAction == Action.GRAB): # Assume GRAB only done if Glitter was present
                self.agentHasGold = True
        if (lastAction == Action.CLIMB):
            pass # do nothing; if CLIMB worked, this won't be executed anyway
        # HW5 requirement 3a
        if percept.glitter:
            self.goldLocation = self.agentLocation
            print("Found gold at " + str(self.goldLocation))
        # HW5 clarification: track world size
        new_max = max(self.agentLocation[0], self.agentLocation[1])
        if new_max > self.worldSize:
            self.worldSize = new_max
        # HW5 requirement 3b
        if not game_over:
            self.UpdateSafeLocations(self.agentLocation)
        # HW5 requirement 3c
        if self.agentLocation not in self.visitedLocations:
            self.visitedLocations.append(self.agentLocation)
    
    def UpdateSafeLocations(self, location):
        # HW5 requirement 3b, and HW5 clarification about not known to be unsafe locations
        # Add current and adjacent locations to safe locations, unless known to be unsafe.
        if location not in self.safeLocations:
            self.safeLocations.append(location)
            self.searchEngine.AddSafeLocation(location[0], location[1])
        for adj_loc in self.AdjacentLocations(location):
            if (adj_loc not in self.safeLocations) and (adj_loc not in self.unsafeLocations):
                self.safeLocations.append(adj_loc)
                self.searchEngine.AddSafeLocation(adj_loc[0], adj_loc[1])
        
    def SafeUnvisitedLocation(self):
        # Find and return safe unvisited location
        for location in self.safeLocations:
            if location not in self.visitedLocations:
                return location
        return None
    
    def RemoveOutsideLocations(self):
        # Know exact world size, so remove locations outside the world.
        boundary = self.worldSize + 1
        for i in range(1,boundary):
            if [i,boundary] in self.safeLocations:
                self.safeLocations.remove([i,boundary])
                self.searchEngine.RemoveSafeLocation(i, boundary)
            if [boundary, i] in self.safeLocations:
                self.safeLocations.remove([boundary, i])
                self.searchEngine.RemoveSafeLocation(boundary, i)
        if [boundary, boundary] in self.safeLocations:
            self.safeLocations.remove([boundary, boundary])
            self.searchEngine.RemoveSafeLocation(boundary, boundary)
    
    def AdjacentLocations(self, location):
        # Return list of locations adjacent to given location. One row/col beyond unknown
        # world size is okay. Locations outside the world will be removed later.
        X = location[0]
        Y = location[1]
        adj_locs = []
        if X > 1:
            adj_locs.append([X-1,Y])
        if Y > 1:
            adj_locs.append([X,Y-1])
        if self.worldSizeKnown:
            if (X < self.worldSize):
                adj_locs.append([X+1,Y])
            if (Y < self.worldSize):
                adj_locs.append([X,Y+1])
        else:
            adj_locs.append([X+1,Y])
            adj_locs.append([X,Y+1])
        return adj_locs
